### ISSUE TITLE
Proposal: Stages

### ISSUE THREAD

**Issue by krancour** [kubernetes/helm#2243]
Hooks may not be sufficient to cover some common use cases. I want to propose something new...

I use Rails for various hobby projects. I've been thinking a lot about how I can run database migrations during a `helm install` or `helm upgrade`. The obvious answer is to put these in a job. Things get tricky very fast...

Rails learns the schema of model objects by inspecting the database. That means that if a pod serving web traffic comes up live and ready while migrations are incomplete, that web-serving process is working off a partial schema...

So I've thought about moving migration jobs into a `pre-install` and `pre-upgrade` hook. Two new problems pop up right away...

1. To have a _new_ migration job run each time a `helm upgrade` occurs, I have to append a unique numeric value to the end of the job name. This means completed jobs accumulate over time. I'm sure there are ways I can work around this, but they all seem like a lot of work. #1769 touches on this. But there's also a more serious problem...

2. The migration needs sensitive information-- like the database password. Naturally this is stored in a secret. The pod running the web process needs this same secret. For the migration to have access to it, it should be in the `pre-install` hook. But things in hook aren't managed by Helm on an ongoing basis. So this means this secret can never be updated.

One workaround for the above could be to have a secret used by the migration, and as with the migration itself, that secret can have a name that changes with each release. The web process can have its own secret with a normal (static/predictable) name that doesn't change with every release. Except for the name, it would be no different from the secret utilized by the migration. That downside to this is that not only will migrations accumulate in k8s, but so will one-off secrets that were used by those migrations... yuck.

I want to propose a cleaner way to handle these sort of scenarios-- "stages". I'm borrowing the idea from Puppet. The basic idea would be that, unlike with hooks, things in _every_ stage would be fully managed by Helm, but the stages would represent an explicit order in which resources should be added, updated, or deleted.

Take my migration example:

- Stage 10: Add/upgrade the secret
- Stage 20: Add a migration job (I don't think jobs can really be updated, so refer back to #1769 for further discussion on cleaning up completed jobs)
- Stage 30: Add/upgrade all other resources, some of which (like the deployment) depend explicitly on the secret in stage 10 and implicitly on the migration job in stage 20 having completed.

**Comment by thomastaylor312** [comment]
@krancour Thanks for the proposal! Would #2157 at least partially address the issue?

**Comment by krancour** [comment]
@thomastaylor312 nope. I am already using weighted hooks. They are useful for certain things, but not for solving the problems I have described.

Fundamentally what I want is for "hooks" to start managing the lifecycle of their resources. I recognize that's a breaking change, so instead I'm proposing "stages" as a new, fully-managed, alternative to hooks.

**Comment by technosophos** [comment]
I've used many DB migration tools that follow this pattern, but it's much more complicated in this situation. That reason is that Helm does not have any information about what is being altered by each stage. Thus, `helm rollback` is highly likely to break, as are cases where things like image tags are changed out of band.

Example:

- I intall my rails app. Stage = 1
- I upgrade my rails app, with a new stage. Stage = 2
- I upgrade my rails app again, with no stage changes
- I rollback... Under what conditions to I unstage?

It seems to me that the conditions that would trigger an unstaging are not clear because we don't know what the staging and unstaging involves.

**Comment by krancour** [comment]
@technosophos I think you've lost me.

Although my use case is db migrations (and the things on which they depend-- e.g. a secret), and although I think it's a common use case, let's not get hung up on the particulars of that.

The fundamental difficulty with hooks, as I see it, is that resources (of any kind) defined in a hook are not managed by Helm. This leads people to get creative so that, for instance, each time a `pre-upgrade` hook is run, it's not trying to create create resources that already exist. This could lead to appending unique or random numbers to the end of resource names. In turn, this could lead to an accumulation of resources that are unmanaged and unneeded. The solution... create a `post-upgrade` job to do the cleanup? But that job will itself be an unmanaged resource. It will have to clean up itself?

I think what I am really driving at here is that creating _more_ unmanaged resources to clean up all the other unmanaged resources just seems like an unsatisfyingly complex way to solve for Helm not managing resources we might want managed.

Ultimately, what I was proposing is some optional hook-like thing that allows the resources defined within to still me managed.

**Comment by technosophos** [comment]
Oh. THAT I would support wholeheartedly. Hooks have come to serve double-duty as (a) a way to create unmanaged stuff, and (b) a way to do simple workflow. But really, hooks aren't very good at (b).

I could see doing what you propose here, and creating an alternative to hooks. Or I could see adding another annotation that would indicate that a thing should be cleaned up. That might be as simple as something like:

```yaml
annotations:
  "helm.sh/hookCleanupPolicy": delete
```

Policies could be:

- ignore (default): Leave thing unmanaged
- delete: Delete at the end of the hook run
- replace: Leave it there, but replace it each upgrade

I'm open to whatever option is best for the user.

**Comment by krancour** [comment]
> Hooks have come to serve double-duty as (a) a way to create unmanaged stuff, and (b) a way to do simple workflow. But really, hooks aren't very good at (b).

Are a lot of people really wanting to do a.? (Merely curious.)

> Or I could see adding another annotation that would indicate that a thing should be cleaned up.

That would be incredibly useful!

"Stages" or something like it could _still_ be useful to me. Going back to my original example, migration jobs _and_ web containers mutually depend on a common secret. That means I'd like the secret both managed _and_ created/updated before the migration job.

I'd want this:

- Stage 10: Create / update managed secret
- Stage 20: Create job; delete after success
- Stage 30: Everything else

Like you, I'm open to whatever options most reasonably make a workflow like the above possible.

**Comment by technosophos** [comment]
In regards to (a), I see it basically used to create TPRs, secrets, and configmaps. So, yeah, it gets used that way. But honestly I get many more questions about supporting (b).

I'm having some crazy ideas about how we might be able to build the stages thing as an add-on to Helm. The difficulty with the idea is that we want to sorta "script" something for each stage, and that's not currently easy to do. But I think we _could_ do it.

**Comment by technosophos** [comment]
Okay, I talked to @krancour in person, and for my own benefit am going to try to re-explain what he said in my own words:

The problem to solve:
I have a Rails app. I want to have it (a) create a config map, (b) run a migration (as a Job) that uses that config map, (c) start a deployment that uses that config map.

And I want to be able to run an upgrade cycle that repeats (b) and (c), but doesn't leave around a giant pile of completed jobs.

There are two parts to a solution:

1. The easy one is that we should make it possible to clean up jobs that are completed. We might be able to do this by supporting an annotation on the Job.
2. The slightly less easy one is that we would want to implement the ordering/waiting logic of Hooks, but inside of the main installation run. This is what krancour calls "stages" above -- basically integrating hook weights and the wait logic into the main release process. So _any_ resource in the release (not just hooks) could have weights. And the install/upgrade command would simply run these incrementally until there were no more stages to run.

**Comment by krancour** [comment]
@technosophos good summary. One amendment:

> And I want to be able to run an upgrade cycle that repeats (b) and (c), but doesn't leave around a giant pile of completed jobs.

I'd want (a) to repeat, too. The config map may have new k/v's upon `helm upgrade`.

Spot on on all other counts. Thanks for taking the time to chat!

**Comment by thomastaylor312** [comment]
@technosophos Has this been hammered out enough to label as a feature and drop in a milestone?

**Comment by wstrange** [comment]
I think this is related to https://github.com/kubernetes/helm/issues/2207

**Comment by thomastaylor312** [comment]
@wstrange I think there probably is some overlap here.

**Comment by spiddy** [comment]
I'm adding my use case for reference also:

* I have a meta-chart A that depends on charts B, C and D.
* C & D needs B to be running before starting up.

I want to be able to do "helm install A" and install everything without having to do ordering outside of helm, something like fire-and-forget kinda thing (or call it declarative deployment)

So I think Stages makes sense for this case also, taking into account the dependencies ordering also.

Not sure if there is another way to do this currently

**Comment by thomastaylor312** [comment]
@technosophos Is this at a point when we feel safe labeling this as a feature or is more discussion needed?

**Comment by bacongobbler** [comment]
I wonder if this is something we should consider out of helm's scope and instead turn users to tooling like [brigade](http://brigade.sh/) to handle these complex use cases. @technosophos any further opinion on this?

**Comment by thomastaylor312** [comment]
@bacongobbler That would actually be probably the best idea. I know we have talked about trying to make sure that Helm doesn't become a full lifecycle management tool over a package manager

**Comment by bacongobbler** [comment]
Cool. I'm going to aggressively close this as something we will not implement in Helm, but we can re-open this for further discussion if there's disagreement on Helm's scope. Thanks everyone!

**Comment by technosophos** [comment]
After some discussion with @krancour in Slack, I am re-opening this. It's still a relevant problem we need to solve.

**Comment by hobti01** [comment]
@technosophos, is this feature intended to handle the use case of @spiddy and @wstrange or should we write a custom rudder implementation as @bacongobbler suggests (could you please provide a doc link to write our own rudder back end)?

I would expect that proper ordering of packages (Charts) during install (and uninstall) is the responsibility of the package manager.  We're facing some difficulties at the moment when databases (child charts) are uninstalled out from under apps that the create crash looping StatefulSet Pods that don't delete nicely.  Simply installing dependencies first and uninstalling them last would relieve this.

**Comment by bacongobbler** [comment]
Where did I suggest implementing a custom rudder? :)

If you're referring to https://github.com/kubernetes/helm/issues/2243#issuecomment-340919277, I was suggesting a completely different tool called [brigade](https://github.com/Azure/brigade) which can be [used in pair with helm](https://github.com/technosophos/hello-helm/blob/7b8b7fb4dd0d9cb31b7e789b8874800b2a80ae31/brigade.js#L22-L27) to handle this use case of running database migrations before a `helm upgrade`.

**Comment by hobti01** [comment]
I didn't mean to misquote you, but I read comment https://github.com/kubernetes/helm/issues/2207#issuecomment-340916264  as "use brigade or write a rudder back end" apologies if I mis-understood :-)

Brigade seems a bit heavy handed for Chart sequencing, although I could see how one could install charts one by one in an order.  But then Chart dependencies in helm become superfluous. ;-)

I'd like to understand the design direction before contemplating PRs for an undesired feature. :-)

**Comment by krancour** [comment]
@hobti01 this issue doesn't have to do with the install / uninstall sequence of dependent charts. It has to do with a sequence _within_ a given chart. The canonical example I was working from was a job for db migrations may need to run (to completion) before a deployment is created or updated. And secrets and configmaps common to those jobs may be common to both the migration job and the deployment and therefor must exist before either. This ends up requiring the following ordering:

secrets --> migration job --> deployment

**Comment by zuzzas** [comment]
Would help us immensely as well. @krancour's use case is the most requested one, but we've also had a number of cases with advanced database migrations. That feature seems like a massive undertaking to my inexperienced eyes, but it'll help a lot.

**Comment by amalaowala** [comment]
+1 
facing similar issue here. I'm in the process of migrating our kubernetes deployment to helm and have hit the same db init/migrate scenario. Our services cannot be deployed until their respective DBs have been initialized and had migrations run. 

Unfortunately our DB migrations job require the same set of configmaps & secrets to connect to the DB that our services need.

Ideal scenario

Pre-Install Hook 0 : Install necessary configmaps + secrets
Pre-Install Hook 10 : Run DB migrations for services
Install stage : install + update all services

Until this is available in Helm, is there a best practices pattern for dealing with this scenario if we want to use Helm?

**Comment by wfhartford** [comment]
@amalaowala I don't know about best practices, but I've had good luck solving a very similar problem as follows:

My chart includes a Job which performs the database install or upgrade. This is not part of a pre-install hook, so configmaps and secrets are installed before it. When this job completes, it updates or adds an annotation on the database service with a value which identifies the state to which the database has been updated. This job has to be named uniquely for each database state so that it gets deleted and recreated by tiller on each upgrade.

My application deployments are part of the same chart and deployed at about the same time as the job. These application deployments have an init container which waits for the database service to be annotated with the expected value.

This works nicely for me because on an upgrade, the previous version of the application continues to run while the database is being upgraded. When the database is upgraded, the init container completes successfully, allowing the new version of the application to start, once it is ready, the old version of the application is stopped.

**Comment by pv93** [comment]
@wfhartford Why not run the migration inside of the initContainer? Once it exits (successfully) your app can kick off knowing that the migrations ran.

**Comment by wfhartford** [comment]
@pv93 That's a good point, and actually a path I headed down at one point and forgot about until your comment. It would probably work for most situations, but I had 4 containers that had to wait for one database migration to complete.

**Comment by davidemerritt** [comment]
@wfhartford - Thanks for providing the blocking solution for pre-deployment steps as you described above with: 

`When this job completes, it updates or adds an annotation on the database service with a value which identifies the state to which the database has been updated.`

I am still new to kubernetes/helm and was wondering what mechanism you used for both editing/adding the annotation on the service from within the job, as well as the mechanism you used for your initContainer waiting. Not sure how to achieve this within the cluster. Thanks

**Comment by kycrow32** [comment]
This proposal sought to address a very common use case:  preparing or upgrading a database during a helm install.

A couple of workarounds were proposed.
@wfhartford suggested having a job add an annotation to a kubernetes resource (the database service).
In production environments, permissions may be such that a Job is not allowed to do this.

@bacongobbler suggested using a tool like brigade.
That would require duplicating the templates used to define the database credentials, which violates DRY.

I know it will remain out of scope for helm, but I will comment here anyway.
Helm needs some way to address this basic use case, that does not involve extra permissions, and does not require duplicating templates/resources/credentials.
If there is such a solution, someone make me look stupid.

If there is not, then helm should provide customizable dependencies among resources, so that I can say "MyDeployment depends on MyJob", and "MyJob depends on MyConfigMap".
The default could remain the current lifecycle, where "MyDeployment" defaults to depending on all ConfigMaps (or anything else before it in the helm lifecycle).
